pkg/constraints/constraint.go

```diff
@@ -0,0 +1,83 @@
+package constraints
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+)
+
+// OLMConstraintType is the schema "type" key for all constraints known to OLM
+// (except for legacy types).
+const OLMConstraintType = "olm.constraint"
+
+// Constraint holds parsed, potentially nested dependency constraints.
+type Constraint struct {
+	// Constraint message that surfaces in resolution
+	// This field is optional
+	Message string `json:"message"`
+
+	// The cel struct that contraints CEL expression
+	Cel *Cel `json:"cel,omitempty" yaml:"cel,omitempty"`
+
+	// Package defines a constraint for a package within a version range.
+	Package *PackageConstraint `json:"package,omitempty" yaml:"package,omitempty"`
+	// GVK defines a constraint for a GVK.
+	GVK *GVKConstraint `json:"gvk,omitempty" yaml:"gvk,omitempty"`
+
+	// All, Any, and None are compound constraints. See this enhancement for details:
+	// https://github.com/operator-framework/enhancements/blob/master/enhancements/compound-bundle-constraints.md
+	All *CompoundConstraint `json:"all,omitempty" yaml:"all,omitempty"`
+	Any *CompoundConstraint `json:"any,omitempty" yaml:"any,omitempty"`
+	// A note on None: this constraint is not particularly useful by itself.
+	// It should be used within an All constraint alongside some other constraint type
+	// since saying "none of these GVKs/packages/etc." without an alternative doesn't make sense.
+	None *CompoundConstraint `json:"none,omitempty" yaml:"none,omitempty"`
+}
+
+// CompoundConstraint holds a list of potentially nested constraints
+// over which a boolean operation is applied.
+type CompoundConstraint struct {
+	Constraints []Constraint `json:"constraints"`
+}
+
+// GVKConstraint defines a GVK constraint.
+type GVKConstraint struct {
+	Group   string `json:"group"`
+	Kind    string `json:"kind"`
+	Version string `json:"version"`
+}
+
+// PackageConstraint defines a package constraint.
+type PackageConstraint struct {
+	// Name of the package.
+	Name string `json:"name"`
```

Comment:
I'm not sure if the package constraint has changed but this doesn't look correct to me. It is supposed to be `packageName` and `Version`? Do we unmarshal the JSON blob directly into this `PackageConstraint` struct or this is struct is some sort of internal representation?

Reply:
It was `packageName` on `olm.package.required` but this is a completely different type so the name doesn't have to match. `packageName` is redundant anyway.

Reply:
I'm slightly confused. So this `PackageConstraint` is a new type of constraint? Or you are saying that if you want to use compound constraint then you need to use this new `PackageConstraint`, not the existing `olm.package.required`?

Reply:
If you want to use `olm.constraint` then you need to use the new `PackageConstraint` type. The only difference for users is `packageName -> name` (other than the different parent schema).

Reply:
Just to confirm what you were saying @estroz, the `olm.constraint.package` is a more fleshed out alternative to using `olm.package.required` and `olm.package.required` will not be unsupported with this change.

Reply:
So if they want to use solo package dependency, they need to use existing schema but if they want to use compound, they need to use the new scheme even though they are the same but just different field names. I’m not entirely sure what the purpose of this is? It is an unnecessary introduction of a new schema for the sake of compound constraints even they are basically the same. If we continue to use the existing schema for compound it should still work.

Reply:
Yep!

Reply:
I'm not tied to name over packageName, but that is [what is approved](https://github.com/operator-framework/enhancements/blob/master/enhancements/compound-bundle-constraints.md#implementation-detailsnotesconstraints) in the EP.